metrics/README.md

```diff
@@ -0,0 +1,103 @@
+access to (Google) application default credentials. It's not clear that we want
+to build all of those backends into the core of `knative.dev/pkg` and all
+downstream dependents, and we'd like all the code shipped in `knative.dev/pkg`
+to be able to be tested without needing any special environment setup.
+
+With the current direct-integration setup, there needs to be initial and ongoing
+work in `pkg` (which should be high-value, low-churn code) to maintain and
+update stats exporters which need to be statically linked into ~all Knative
+binaries. This setup also causes problems for vendors who may want or need to
+perform an out-of-tree integration (e.g. proprietary or partially-proprietary
+monitoring stacks).
+
+Another problem is that each vendor's exporter requires different parameters,
+supplied as Golang `Options` methods which may require complex connections with
+the Knative ConfigMap. Two examples of this are secrets like API keys and the
+Prometheus monitoring port (which requires additional service/etc wiring).
+
+See also
+[this doc](https://docs.google.com/document/d/1t-aov3XrhobjCKW4kwScY44QAoahiwxoyXXFtZyL8jw/edit),
+where the plan was worked out.
+
+## The plan
+
+OpenCensus (and eventually OpenTelemetry) offers an sidecar or host-level agent
+with speaks the OpenCensus protocol and can proxy from this protocol to multiple
+backends.
+
+![OpenCensus Agent configuration](https://github.com/census-instrumentation/opencensus-service/raw/master/images/opencensus-service-deployment-models.png)
+(From OpenCensus Documentation)
+
+**We will standardize on export to the OpenCensus export protocol, and encourage
+vendors to implement their own OpenCensus Agent or Collector DaemonSet, Sidecar,
+or other
+[OpenCensus Protocol](https://github.com/census-instrumentation/opencensus-proto/tree/master/src/opencensus/proto/agent)
+service which connects to their desired monitoring environment.** For now, we
+will use the `config-observability` ConfigMap to provide the OpenCensus
+endpoint, but we will work with the OpenTelemetry group to define a
+kubernetes-friendly standard export path.
+
+**Additionally, once OpenTelemetry agent is stable, we will propose adding the
+OpenTelemetry agent running on a localhost port as part of the runtime
+contract.**
+
+We need to make sure that the OpenCensus library does not block, fail, or queue
+metrics in-process excessively in the case where the OpenCensus Agent is not
+present on the cluster. This will allow us to ship Knative components which
+attempt to reach out the Agent if present, and which simply retain local
+statistics for a short period of time if not.
+
+### Concerns
+
+- Unsure about the stability of the OpenCensus Agent (or successor). We're
+  currently investigating this, but the OpenCensus agent seems to have been
+  recommended by several others.
+- Running `fluentd` as a sidecar was very big (400MB) and had a large impact on
```

Comment:
Why do you mention `fluentd` here? I think you mean `OpenCensus agent`?

Reply:
I mention `fluentd` here because at one point we ran `fluentd` as a sidecar, rather than as a DaemonSet. We switch to the DaemonSet due to the problems mentioned.

Running as a sidecar has the following benefits:

* Resource usage is properly accounted to the user's namespace
* Credentials are similarly associated with the user's namespace
* It's easier to configure `fluentd` to export for exactly container than to pick up all the containers on the node.

The drawbacks of running as a DaemonSet with our historical approach:

* Upgrades become trickier, as they need to roll out through the Deployment/Revision reconciliation which is "perpendicular" to the typical Knative rollout (i.e. it updates the running Pods of a Revision, rather than spawning a new revision).
* Requiring `fluentd` container to be present caused slow starts when the container hadn't already been loaded on the node.
* Running as a DaemonSet is closer to "logging as a service" model that seems long-term desirable, rather than "everyone figures out logging themselves". It also decouples the need for Knative controllers to know about fluent plugins and how to export logs, which simplifies `config-logging` for application workloads considerably.

Reply:
What he means is that a topic that came up in our discussion was the potential for injecting fat sidecars materially affecting cold-start latency.  This was an issue with sidecar-based `fluentd`, and we'd like to understand the impact before encouraging the pattern for metrics.

My hot take is: `fluentd` is a beast, but the pattern would likely have been fine for `fluentbit`, so I'm skeptical this will matter if the metrics sidecar is a lean Go container.

Reply:
Thanks for explaining.